shell/platform/android/io/flutter/embedding/engine/FlutterJNI.java

```diff
@@ -939,6 +942,117 @@ public void setLocalizationPlugin(@Nullable LocalizationPlugin localizationPlugi
+   * symbols to load into the dart VM.
+   *
+   * <p>Successful loading of the dart library completes the future returned by loadLibrary() that
+   * triggered the install/load process.
+   *
+   * @param loadingUnitId The loadingUnitId is assigned during compile time by gen_snapshot and is
+   *                      automatically retrieved when loadLibrary() is called on a dart deferred
+   *                      library. This is used to identify which dart deferred library the resolved
+   *                      correspond to.
+   *
+   * @param searchPaths   An array of paths in which to look for valid dart shared libraries. This
+   *                      supports paths within zipped apks as long as the apks are not compressed
+   *                      using the `path/to/apk.apk!path/inside/apk/lib.so` format. Paths will be
+   *                      tried first to last and ends when a library is sucessfully found. When the
+   *                      found library is invalid, no additional paths will be attempted.
+   */
+  @UiThread
+  public void loadDartDeferredLibrary(
+      int loadingUnitId,
+      @NonNull String[] searchPaths) {
+    ensureRunningOnMainThread();
+    ensureAttachedToNative();
+    nativeLoadDartDeferredLibrary(nativePlatformViewId, loadingUnitId, searchPaths);
+  }
+
+  private native void nativeLoadDartDeferredLibrary(
+      long nativePlatformViewId,
+      int loadingUnitId,
+      @NonNull String[] searchPaths);
+
+  /**
+   * Specifies a new AssetManager that has access to the dynamic feature's assets in addition to the
+   * base module's assets.
+   *
+   * @param assetManager    An android AssetManager that is able to access the newly downloaded assets.
+   *
+   * @param assetBundlePath The subdirectory that the flutter assets are stored in. The typical value
+   *                        is `flutter_assets`.
+   */
+  @UiThread
+  public void updateAssetManager(
+      @NonNull AssetManager assetManager, @NonNull String assetBundlePath) {
+    ensureRunningOnMainThread();
+    ensureAttachedToNative();
+    nativeUpdateAssetManager(nativePlatformViewId, assetManager, assetBundlePath);
+  }
+
+  private native void nativeUpdateAssetManager(
+      long nativePlatformViewId,
+      @NonNull AssetManager assetManager,
+      @NonNull String assetBundlePath);
+
+  /**
+   * Indicates that a failure was encountered during the Android portion of downloading a dynamic
+   * feature module and loading a dart deferred library, which is typically done by
+   * DynamicFeatureManager.
+   *
+   * <p>This will inform dart that the future returned by loadLibrary() should complete with an error.
+   *
+   * @param loadingUnitId The loadingUnitId that corresponds to the dart deferred library that
+   *                      failed to install.
+   *
+   * @param moduleName    The name of the module that includes the loadingUnitId. This parameter is
```

Comment:
I'm somewhat expecting the loadDartDeferredLibrary and this signature to be pretty close since one signals success and one signals failure. 

e.g. what happens if it fails not while downloading but while searching for the .so? If you have the module name while downloading, why not have the module name while dlopening?

Reply:
The moduleName param here is just used for logging the name for convenience. I think I'll just remove this parameter as it is not necessary at all to signal failure to the rest of the code stack.